File: app/src/main/java/com/blog/demo/image/path/PathNormalActivity.java
package com.blog.demo.image.path;

import android.app.Activity;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

public class PathNormalActivity extends Activity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(new PathNormalView(this));
    }

    class PathNormalView extends View {

        private Paint mRedPaint;

        public PathNormalView(Context context) {
            this(context, null);
        }

        public PathNormalView(Context context, @Nullable AttributeSet attrs) {
            super(context, attrs);

            mRedPaint = new Paint();
            mRedPaint.setAntiAlias(true);
            mRedPaint.setColor(Color.RED);
            mRedPaint.setStyle(Paint.Style.STROKE);
            mRedPaint.setStrokeWidth(5);
        }

        @Override
        protected void onDraw(Canvas canvas) {
            Path path = new Path();
            path.moveTo(50, 50);
            path.lineTo(200, 100);
            path.rMoveTo(50, 0);
            path.rLineTo(100, 100);

            path.addRect(50, 250, 300, 350, Path.Direction.CW);
            path.addRoundRect(50, 400, 300, 500, 20, 10, Path.Direction.CW);
            path.addRoundRect(400, 400, 650, 500, new float[]{10, 20, 10, 20, 20, 10, 20, 10},
                    Path.Direction.CW);

            path.addOval(50, 550, 300, 700, Path.Direction.CW);
            path.addCircle(525, 625, 75, Path.Direction.CW);

            path.addArc(50, 750, 300, 900, 0, 180);
            path.addArc(400, 750, 650, 900, 90, 225);

            path.arcTo(50, 950, 300, 1100, 0, 180, true);
            path.arcTo(400, 950, 650, 1100, 90, 225, false);
            canvas.drawPath(path, mRedPaint);
        }
    }

}

File: app/src/main/java/com/blog/demo/image/path/PathNormalActivity.kt
package com.blog.demo.image.path

import android.app.Activity
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.os.Bundle
import android.util.AttributeSet
import android.view.View

class PathNormalActivity : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(PathNormalView(this))
    }

    private inner class PathNormalView(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {

        private val mRedPaint: Paint = Paint()

        init {
            mRedPaint.isAntiAlias = true
            mRedPaint.color = Color.RED
            mRedPaint.style = Paint.Style.STROKE
            mRedPaint.strokeWidth = 5f
        }

        override fun onDraw(canvas: Canvas) {
            val path = Path()
            path.moveTo(50f, 50f)
            path.lineTo(200f, 100f)

            path.rMoveTo(50f, 0f)
            path.rLineTo(100f, 100f)

            path.addRect(50f, 250f, 300f, 350f, Path.Direction.CW)

            path.addRoundRect(50f, 400f, 300f, 500f, 20f, 10f, Path.Direction.CW)
            path.addRoundRect(400f, 400f, 650f, 500f,
                floatArrayOf(10f, 20f, 10f, 20f, 20f, 10f, 20f, 10f), Path.Direction.CW)

            path.addOval(50f, 550f, 300f, 700f, Path.Direction.CW)
            path.addCircle(525f, 625f, 75f, Path.Direction.CW)

            path.addArc(50f, 750f, 300f, 900f, 0f, 180f)
            path.addArc(400f, 750f, 650f, 900f, 90f, 225f)

            path.arcTo(50f, 950f, 300f, 1100f, 0f, 180f, true)
            path.arcTo(400f, 950f, 650f, 1100f, 90f, 225f, false)

            canvas.drawPath(path, mRedPaint)
        }

    }

}